src/AsyncEventSource.h

```diff
@@ -53,58 +60,150 @@ class AsyncEventSourceResponse;
 class AsyncEventSourceClient;
 using ArEventHandlerFunction = std::function<void(AsyncEventSourceClient* client)>;
 using ArAuthorizeConnectHandler = ArAuthorizeFunction;
+// shared message object container
+using AsyncEvent_SharedData_t = std::shared_ptr<std::string>;
 
+/**
+ * @brief Async Event Message container with shared message content data
+ *
+ */
 class AsyncEventSourceMessage {
+
   private:
-    uint8_t* _data;
-    size_t _len;
-    size_t _sent;
-    // size_t _ack;
-    size_t _acked;
+    const AsyncEvent_SharedData_t _data;
+    size_t _sent{0};  // num of bytes already sent
+    size_t _acked{0}; // num of bytes acked
 
   public:
-    AsyncEventSourceMessage(const char* data, size_t len);
-    ~AsyncEventSourceMessage();
+    AsyncEventSourceMessage(AsyncEvent_SharedData_t data) : _data(data) {};
```

Comment:
`AsyncEventSourceMessage` class is not used directly by user code, user code interacts with `AsyncEventSource` and `AsyncEventSourceResponse` classes. And `AsyncEventSource` does not have any methods accepting `std::string` objects.
But as for me it's actually _not good_ at all, I would rather add `std::strings` there in some near future. With both 8266 and esp32 supporting C++11 for ages (and even c++23 in a recent versions) sticking solely to Arduino `String`s is just dragging a huge stone from 8-bit era on your neck while trying to fly with WiFi :) `ArduinoJson` which is almost always used together with AsyncWebServer has full support for `std::string` and even `std::string_view` out of the box.

As for move constructor for `String` it's just a wrapper, not a real move semantics implementation.
If you check the source for `WString.cpp` it uses new buffer, mem allocation and `memmove` under the hood, which is a kind if `memchr()` wrapper. So it does not moves actually, it copies with only difference that memory ranges for source and destination could overlap partially.
Without real move semantics this PR does not makes much sense, it would be simpler to keep using pure C's `uint8_t*`,  `malloc()`  and `strcpy`.

maybe just park this PR for 4.x branch milestone?

Reply:
Park or switch to String considering that in 4.x, everything will be moved to std::string ? That could be an option also. Because of 4.x only supports ESP32, we will probably also use std::string everywhere ?

Reply:
Or as an alternative, I could create now a `v4` branch, we start cleaning it up, we merge this PR with String in `main` and we cherry pick it for `v4` and switchh to std::string. This way the code base being similar will facilitate fixes.

Reply:
I'm all OK with v4 branch for for some significant changes things. But I do not get your point with `String`. You mean merge this PR into v4 branch as-is, but rework it to use `String` and backport to existing v3? Possible, but code base will diverge, so as supporting new features. `std::string` anyway has a bit different set of methods compared to `String`. But that not not much of change, so why not?

Reply:
The current public API uses `String`, so users of ESpAsyncWS are probably anyway using String in their project. 
So adding a new method: ` bool write(AsyncEvent_SharedData_t message) {` should be consistent with the existing. That's all what I am saying ;-) That's not normal to have in an API some methods accessible with String and some with std:string.

That is why I was saying:
- either we switch to String
- either we only keep this PR as-is for v4
- or we do both: we merge this improvement with String and we take the opportunity to create v4, and in v4 we switch to std::string, but not only there, but everywhere.

Reply:
> The current public API uses `String`

But for `AsyncEventSourceMessage` there was never `String`, Mathieu,  it was `const char*` _only_!
And if anyone was using `AsyncEventSourceMessage` with `String`, it could only be instantiated via copy from `string.c_ctr()`, but not from `String&`

here was the old c-tor
```
AsyncEventSourceMessage::AsyncEventSourceMessage(const char* data, size_t len)
    : _data(nullptr), _len(len), _sent(0), _acked(0) {
  _data = (uint8_t*)malloc(_len + 1);
  if (_data == nullptr) {
    _len = 0;
  } else {
    memcpy(_data, data, len);
    _data[_len] = 0;
  }
}
```

I've kept the same `const char*` overload, just added a c-tor with `shared_pointer`
```
  public:
     AsyncEventSourceMessage(const char* data, size_t len);  // same as before
+    AsyncEventSourceMessage(AsyncEvent_SharedData_t data) : _data(data) {}; // added
```

if you mean that a constructor from `String` is needed than it's just a matter of adding a constructor overload
```
+    AsyncEventSourceMessage(const String& data) : AsyncEventSourceMessage(data.c_str()) {};
```

Same with `AsyncEventSourceClient::write` - it was _only_ `write(const char* message, size_t len)`, no `String` overload existed.


If it's the question of NOT adding `AsyncEvent_SharedData_t` whatever type it hides behind then it does not matter `String` or `std::string` we use, shallow copy won't be possible.

If you mean that `AsyncEvent_SharedData_t` should be
```c++
using AsyncEvent_SharedData_t = std::shared_ptr<String>;
```

then I'm not sure this change would make much difference to current implementation with `malloc` because `String`'s move semantics implementation is questionable at least. Still could benefit from `std::shared_ptr` though.

IMHO, let's postpone this to 4.x and skip this `String` part, otherwise we would introduce a change in current implementation and then introduce a new change again in 4.x. For the sake of stability and testing efforts I do not see a reason to go through additional `String` phase and have two major changes except of one.
Or let's agree to stick to Arduino's way and not to touch `std::string` containers at all, then we can bring it now with `String` and skip `std::string` part later on till 5.x or 6.x times maybe :)

Reply:
hello,

> Same with AsyncEventSourceClient::write - it was only write(const char* message, size_t len), no String overload existed.

I know about the overload. It does not change the fact that I find weird introducing this new method in v3 where all the API is String-based (and I completely understand that my vision is debatable ;-) ) because it is not consistent. It much more looks like a patchy thing when not knowing what it does.

A user could even wonder why this type is there for ? Should he use it ? etc. It's not about String vs std::string, it is just about api consistency. Why would we introduce string in write() ? why there is no `String` overload ? why all the other methods do not have std::string overloads ? Etc.

Everything is fine, otherwise. And I know there is no impact and this is backward compatible.

**My question is more related to : should we introduce new API which are `std::string`-based without any `String`-base overload in an API that is currently mostly `String`-based (even if recently we did some refactoring to change that)**

Because in v4, I suspect we will remove any `String` usages as much as possible right, only required and mostly relying on `const char*` also right ? and we'll use std::string as holder classes ? So v4 will be an API focusing on constructing char* parameter types, eventually using std::string for holder classes right ? I suspect that in v4, we won't gave anymore methods using `String` ? Or you were thinking about supporting both String and std::string ?

Reply:
ugh... OK, if you put it in this way... :) Do not have an easy answer for that. Well... I do actually... modern c++ way is to use templates. So there should not be some specific std::string or String typed API, any type would do as long as it has proper type trait. It like ArduinoJson's architecture, just create a proper trait for any type feasible. Easy to say, not that easy to do it all.

Other than templates I'm sure it would be that easy to pick just `const char*` + either String or std::string.
Those damn `String`s has one killer feature that makes users addicted to it, it could be implicitly converted from `Int`s and `float`s effectively stringifying arguments. `std::string` does not have this constructs.
Also, as in such case with this PR some templated types might be required to optimize things. Probably we could hide those methods as private and allowed for friend classes only, but who knows... it might useful for users also to get access to this.
I wanted to check for same dedupe possibility for websockets, but a busy for now.

Not sure... let's think it all trough for a while :)

Reply:
Yes! 
And meanwhile this PR gets tested a bit more, I propose we start a v4 branch and start cleaning the code ?
What do you think ?
We can issue 3.x release and at the same time 4.x rc1 r2 ...

Reply:
I'll support that. When I was young, I was obsessed with not using Arduino String as I thought it was all things evil and caused ESPs to fragment memory like a meth'd up monkey. So in my code I replaced everything with std/stl libraries and you know what? didn't make a difference at all! But I still like the fact the code is portable.

Reply:
Same experience. And I think that it is mostly due to the fact that to really benefit of the inner string move semantic, most of the app would need to be written using string and not even char* as parameters because as soon as const char* is used, it puts the responsibility to the caller to pass a c string or eventually build a buffer to call this method 

But in Arduino world I know little libs using std::string. And the projets using std::string are doomed to do some conversion from String -> char* -> string at one point because all the Arduino core API is using the String class anyway.

Although the argument of using string holds when using esp-idf directly, or for the use case of this pr for internal optimisation.

Reply:
IMO, there were two era's of `String`s - one when avr toolchain had no `STL` support, good old days :)
And then instead of steering toward STL the 8266 was born with it's PROGMEM and problems with aligned access to ROM data. For sure none of STL were able to cope with this and `String` hack was the only way to wrap around ROM data in user-understandable way. Then tons of libs were written this way.
I really hope that as soon as 8266 will be left for good it could give another chance for STL to take the place where it belongs to.
`String`s might be pretty portable and versatile objects for beginners, but at the same time you'll be bound to Arduino env forever and forget about the native SDKs.

Introducing templated methods could allow some sort of freedom and reduce overloading burden, but that might be really challenging for beginners to understand the pro's and con's and deal with compiler errors.
`ArduinoJson` is a really nice example of a software with good design in this regard, but it requires a very good cpp experience to implement and support a strictly designed code.

I like the idea with v4-something... let's see were we could get there and await for community feedbacks.

Cheers!

P.S.  It turned into a nice discussion thread :)))

Reply:
> As for move constructor for `String` it's just a wrapper, not a real move semantics implementation.
> If you check the source for `WString.cpp` it uses new buffer, mem allocation and `memmove` under the hood, which is a kind if `memchr()` wrapper. So it does not moves actually, it copies with only difference that memory ranges for source and destination could overlap partially.
> Without real move semantics this PR does not makes much sense, it would be simpler to keep using pure C's `uint8_t*`, `malloc()` and `strcpy`.

@vortigont : being curious about the move semantic supposedly not correctly implemented in String, I have double-checked and it is actually correctly implemented.

Here is a summary of the key methods:

```c++
String::String(StringSumHelper &&rval) {
  init();
  move(rval);
}

inline void String::init(void) {
  setSSO(false);
  setBuffer(nullptr);
  setCapacity(0);
  setLen(0);
}

void String::move(String &rhs) {
  if (buffer()) {
    if (capacity() >= rhs.len()) {
      memmove(wbuffer(), rhs.buffer(), rhs.length() + 1);
      setLen(rhs.len());
      rhs.invalidate();
      return;
    } else {
      if (!isSSO()) {
        free(wbuffer());
        setBuffer(nullptr);
      }
    }
  }
  if (rhs.isSSO()) {
    setSSO(true);
    memmove(sso.buff, rhs.sso.buff, sizeof(sso.buff));
  } else {
    setSSO(false);
    setBuffer(rhs.wbuffer());
  }
  setCapacity(rhs.capacity());
  setLen(rhs.len());
  rhs.setSSO(false);
  rhs.setCapacity(0);
  rhs.setBuffer(nullptr);
  rhs.setLen(0);
}
```

when doing `class.myString = std::move(parameter)`, the rvalue ctor will make sure `myString` is re-initialized (null buffer). Then, move() is called. The move logic has been extracted in a protected method (which is also used for other use cases) and in case of a big string outside of the scope os SSO optimisation, the code that is executed is:

```c++
  } else {
    setSSO(false);
    setBuffer(rhs.wbuffer());
  }
  setCapacity(rhs.capacity());
  setLen(rhs.len());
  rhs.setSSO(false);
  rhs.setCapacity(0);
  rhs.setBuffer(nullptr);
  rhs.setLen(0);
```

which is exactly what we would expect from a ctor taking rvalue as parameter.

So definitely the PR would work fine with String also.

Reply:
Ywah, I guess it should work then,
but this piece of code looks really strange for me

```c++
void String::move(String &rhs) {
  if (buffer()) {
    if (capacity() >= rhs.len()) {
      memmove(wbuffer(), rhs.buffer(), rhs.length() + 1);
      setLen(rhs.len());
      rhs.invalidate();
      return;
    }
```

if this method is protected then how is this supposed to be called with an existing buffer pointer? And it will preserve larger buffer and copy data... weird... and undocumented... :))) But not needed on new object creation.

I've made a simple sketch to test and added some prints to String constructors

```c++
String a("Some string with some chars here...");
String b("Another one string with some chars here...");

std::string std_a("Some string with some chars here...");
std::string std_b("Another one string with some chars here...");

void setup() {
  Serial.begin(115200);
  Serial.println("Begin test:");
  Serial.print("String a: ");   Serial.println(a.c_str());
  Serial.print("String b: ");   Serial.println(a.c_str());

  Serial.println("\nMoving...");
  String mv_a = std::move(a);
  String mv_b = std::move_if_noexcept(b);


  Serial.print("String a: ");   Serial.println(a.c_str());
  Serial.print("String mv_a: ");   Serial.println(mv_a.c_str());

  Serial.print("String b: ");   Serial.println(b.c_str());
  Serial.print("String mv_b: ");   Serial.println(mv_b.c_str());
```

So here is output... it moves, but does not signify noexept safety
```
Begin test:
String a: Some string with some chars here...
String b: Some string with some chars here...

Moving...
=mv ctor=
=cp ctor=
=cp operator=
String a:
String mv_a: Some string with some chars here...
String b: Another one string with some chars here...
String mv_b: Another one string with some chars here...
```

Reply:
> how is this supposed to be called with an existing buffer pointer?

Yes completely agree! 
My assumption is that they wanted to reuse the allocated space maybe...

But that's right, I saw the same loopholeé

Reply:
> how is this supposed to be called with an existing buffer pointer?

init() is called before to clear the current buffer. So this piece of code is never called from a move ctor.

```c++
String::String(String &&rval) {
  init(); // clear the current buffer
  move(rval);
}
```

So we always go to the move semantic.

and the behaviour with `noexept` is also fine: they call the copy ctor because `noexept` is not used.

So in conclusion you could update this PR to use String ;-)

Also, I found that it will be possible to support both String and std::string in ESPAsync by doing something like this:

```c++
#ifdef ASYNC_USE_STL_STRING
  #include <string>
  #define ASYNC_STR std::string
#else
  #define ASYNC_STR String
#endif
```

- both nearly have the same methods, empty / isEmpty could be replaced by length() which is available for both
- concat calls can be replaced by +=
- conversion calls to int or long could be replaced by `atoi(a_string.c_str())` for exmaple
- etc.

Internally, when not exposed to the user, we can keep std::string or String, but for the API doing like that we should be able to allow a user to switch the API to one or the other.

I tested that yesterday for a big library and it is doable.

Reply:
Adding to this, since you're looking at std::move and strings, have you looked into using `std::string_view` as a cleaner way to get a pass a std::string const reference (e.g. `const &`) without actually creating the string. There's an article on this at https://arduinojson.org/v7/how-to/use-string-view-on-esp32/ and a good explanation [here](https://medium.com/factset/modern-c-in-depth-is-string-view-worth-it-7ae7570b7830) as well. I haven't personally tried it out.

Reply:
ok, I will rework it in a more feasible way.

@proddy `std::string_view`s I use quite frequently in my projects (another reason to steer away from `String`s) and it's a very effective thing for string manipulations in some cases.
But this won't fit here at all for two reasons:
 - it needs C++17, which at least is not default for Arduino Core till IDF 5.3
 - `std::string_view`s are pointers and can't act as a container structs

Reply:
We'll for sure be able to use them in v4 since Arduino 3 will be the target.
Arduino 2 users could add the compile flags to get c++ 17, that's what everyone is doing with platformio.

On my end I only found some few use cases (like  mqtt data callback and for some kind of string search function where strlen is used). But they can be a pain to use because as soon as we need to convert the string or do any operations requiring a normal string anyway we have to transform it, so its a way to move farther away the place where the string copy might be done.

In the case of this project, there is not a lof of places: most of the strings are used as containers...

I tend to keep const char* when:
- underlaying API called takes c string (like most esp-idf API)
- no need for strlen or complex manipulations 
- most of the use case is to pass a string which will be a constant in flash (like: c`onfig.set("foo", std::to_string(50));`